1.process_data/nbconverted/single_cell_qc.py

```diff
@@ -0,0 +1,277 @@
+
+# In[7]:
+
+
+irregular_nuclei_outliers_cdf.sample(n=5, random_state=0)
+
+
+# ### Generate a new dataframe to save the original indices for filtering prior to further preprocessing
+
+# In[8]:
+
+
+# Create a new dataframe for failed QC indices
+failed_qc_indices = irregular_nuclei_outliers[metadata_cols].copy()
+failed_qc_indices["original_index"] = failed_qc_indices.index  # Store original index
+failed_qc_indices = failed_qc_indices.reset_index(drop=True)  # Reset index
+failed_qc_indices["cqc.formfactor_displacement_outlier"] = True
+
+print(failed_qc_indices.shape)
+failed_qc_indices.head()
+
+
+# 2. Nuclei segmentations with holes and also irregular shaped
+# 
+# We need to include this extra condition as it was discovered that there were more poorly segmented cells not caught, especially those over-segmented nuclei that contained holes, which is not common for a nuclei segmentation.
+# 
+# - `Compactness`, which detects irregular nuclei and nuclei containing holes.
+
+# ### Detect outliers and show single-cell image crops with CytoDataFrame (cdf)
+
+# In[9]:
+
+
+# find mis-segmented nuclei due using thresholds that maximizes
+# removing most technical outliers and minimizes good cells
+feature_thresholds = {
+    "Nuclei_AreaShape_Compactness": 4.05,
+}
+
+poor_nuclei_outliers = cosmicqc.find_outliers(
+    df=df_merged_single_cells,
+    metadata_columns=metadata_cols,
+    feature_thresholds=feature_thresholds
+)
+
+poor_nuclei_outliers_cdf = CytoDataFrame(
+    data=pd.DataFrame(poor_nuclei_outliers),
+    data_context_dir=f"/media/jenna/8TB-C/work/JUMP-download-images/JUMP-single-cell/0.download_data/data/images/{plate_id}/orig",
+    data_outline_context_dir=f"/media/jenna/8TB-C/work/JUMP-download-images/JUMP-single-cell/0.download_data/data/images/{plate_id}/outlines",
+    segmentation_file_regex=outline_to_orig_mapping,
+)[
+    [
+        "Nuclei_AreaShape_Compactness",
+        "Image_FileName_OrigDNA",
+    ]
+]
+
+
+print(poor_nuclei_outliers_cdf.shape)
+poor_nuclei_outliers_cdf.sort_values(by="Nuclei_AreaShape_Compactness", ascending=True).head(2)
+
+
+# ### Randomly sample outlier rows to visually inspect if the threshold looks to be optimized
+
+# In[10]:
+
+
+poor_nuclei_outliers_cdf.sample(n=5, random_state=0)
+
+
+# ### Save the original indices for failed single cells to compressed CSV files
+
+# In[11]:
+
+
+# Create a new dataframe for poor nuclei outliers
+poor_qc_indices = poor_nuclei_outliers[metadata_cols].copy()
+poor_qc_indices["original_index"] = poor_qc_indices.index  # Store original index
+poor_qc_indices = poor_qc_indices.reset_index(drop=True)  # Reset index
+poor_qc_indices["cqc.compactness_outlier"] = True
+
+# Merge both outlier dataframes on metadata columns and original_index
+failed_qc_indices = failed_qc_indices.merge(
+    poor_qc_indices,
+    on=[*metadata_cols, "original_index"],
+    how="outer"
+)
```

Comment:
Seeing how this plays out I noticed it might be possible to:
- Define a `yaml` file like the one found within `cosmicqc` [here](https://github.com/WayScience/coSMicQC/blob/main/src/cosmicqc/data/qc_nuclei_thresholds_default.yml) (which includes your threshold sets with the names you've specified manually). 
- Use that `yaml` file to run `cosmicqc.label_outliers(..., feature_thresholds_file=<yaml_file>, ...)`
- Export the file as you need below (also skipping the step to populate the data manually below).

Doing this might lower the complexity involved with the code more generally. It'd also prepare for the threshold sets to be added to `cosmicqc` :) .

Reply:
This is an excellent comment! I am in total agreement, but I think it is out-of-scope for this PR. I think this is great for a refactor, and I will add this as an issue for the repo!

Thank you!